pkg/controllers/state/cluster.go

```diff
@@ -307,15 +314,54 @@ func (c *Cluster) UpdatePod(ctx context.Context, pod *corev1.Pod) error {
 	return err
 }
 
+// MarkPodsSchedulable is called when Karpenter finds that the pod can schedule to a node (theoretical or not)
+// in the scheduling loop, keeping track of the time so it can be used for metrics later.
+func (c *Cluster) MarkPodsSchedulable(pods ...*corev1.Pod) {
+	now := c.clock.Now()
+	for _, pod := range pods {
+		// store the value as now only if it doesn't exist.
+		c.podsSchedulableTimes.LoadOrStore(types.NamespacedName{Name: pod.Name, Namespace: pod.Namespace}, now)
+	}
+}
+
+// PodSchedulableTime returns when Karpenter first thought it could schedule a pod in its scheduling simulation.
+// This returns 0, false if the pod was never considered in scheduling as a pending pod.
+func (c *Cluster) PodSchedulableTime(podKey types.NamespacedName) (time.Time, bool) {
+	val, found := c.podsSchedulableTimes.Load(podKey)
+	if !found {
+		return time.Time{}, false
+	}
+	return val.(time.Time), true
+}
+
+// AckPods marks the pod as acknowledged for scheduling from the provisioner. This is only done once per-pod.
+// If the pod hasn't been previously ACK'd, this emits a metric for the pod.
+func (c *Cluster) AckPods(pods ...*corev1.Pod) {
+	now := c.clock.Now()
+	for _, pod := range pods {
+		// store the value as now only if it doesn't exist.
+		if _, alreadySet := c.podAcks.LoadOrStore(types.NamespacedName{Name: pod.Name, Namespace: pod.Namespace}, true); !alreadySet {
+			timeStart := lo.Ternary(c.initTime.Compare(pod.CreationTimestamp.Time) >= 0, c.initTime, pod.CreationTimestamp.Time)
```

Comment:
I'd question if we should even log a metric based off of the start time of our pods -- I think that's going to drastically skew our metrics on restart

Reply:
My thought was that the metrics could just be a function of the container, but I think actually it's probably better for it to be more consistent across restarts.